pkg/resources/management.cattle.io/v3/globalrolebinding/mutator.go

```diff
@@ -0,0 +1,100 @@
+	"k8s.io/utils/trace"
+)
+
+// Mutator implements admission.MutatingAdmissionWebhook.
+type Mutator struct {
+	globalRoles v3.GlobalRoleCache
+}
+
+// NewMutator returns a new mutator for GlobalRoleBindings.
+func NewMutator(grCache v3.GlobalRoleCache) *Mutator {
+	return &Mutator{
+		globalRoles: grCache,
+	}
+}
+
+// GVR returns the GroupVersionKind for this CRD.
+func (m *Mutator) GVR() schema.GroupVersionResource {
+	return gvr
+}
+
+// Operations returns list of operations handled by this mutator.
+func (m *Mutator) Operations() []admissionregistrationv1.OperationType {
+	return []admissionregistrationv1.OperationType{admissionregistrationv1.Create}
+}
+
+// MutatingWebhook returns the MutatingWebhook used for this CRD.
+func (m *Mutator) MutatingWebhook(clientConfig admissionregistrationv1.WebhookClientConfig) []admissionregistrationv1.MutatingWebhook {
+	mutatingWebhook := admission.NewDefaultMutatingWebhook(m, clientConfig, admissionregistrationv1.ClusterScope, m.Operations())
+	return []admissionregistrationv1.MutatingWebhook{*mutatingWebhook}
+}
+
+// Admit handles the webhook admission request sent to this webhook.
+// If this function is called without NewMutator(..) calls will panic.
+func (m *Mutator) Admit(request *admission.Request) (*admissionv1.AdmissionResponse, error) {
+	listTrace := trace.New("GlobalRoleBinding Mutator Admit", trace.Field{Key: "user", Value: request.UserInfo.Username})
+	defer listTrace.LogIfLong(admission.SlowTraceDuration)
+
+	newGRB, err := objectsv3.GlobalRoleBindingFromRequest(&request.AdmissionRequest)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get %s from request: %w", gvr.Resource, err)
+	}
+
+	err = m.addOwnerReference(newGRB)
+	if err != nil {
+		if errors.As(err, admission.Ptr(new(field.Error))) {
+			return admission.ResponseBadRequest(err.Error()), nil
+		}
+		return nil, fmt.Errorf("failed to add owner reference: %w", err)
+	}
+
+	response := &admissionv1.AdmissionResponse{}
+	if err := patch.CreatePatch(request.Object.Raw, newGRB, response); err != nil {
+		return nil, fmt.Errorf("failed to create patch: %w", err)
+	}
+	response.Allowed = true
+	return response, nil
+}
+
+// addOwnerReference ensures that a GlobalRoleBinding will be deleted when the role it references is deleted.
+func (m *Mutator) addOwnerReference(newGRB *apisv3.GlobalRoleBinding) error {
+	globalRole, err := m.globalRoles.Get(newGRB.GlobalRoleName)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return field.NotFound(field.NewPath("globalrolebinding", "globalRoleName"), newGRB.Name)
+		}
+		return fmt.Errorf("failed to get referenced globalRole: %w", err)
+	}
+	ownerReference := v1.OwnerReference{
+		APIVersion: globalRole.APIVersion,
+		Kind:       globalRole.Kind,
+		Name:       globalRole.Name,
+		UID:        globalRole.UID,
+	}
+	newRefStr := ownerReference.String()
+	for i := range newGRB.OwnerReferences {
+		if newGRB.OwnerReferences[i].String() == newRefStr {
```

Comment:
I'm a bit uncomfortable on relying on `.String()` producing a value suitable for equality checks. Put differently, how confident are we that `a.String() == b.String()` iff `a == b` here? If you have some examples from other projects using this that would probably help.

Reply:
Not tied to the usage of `String()`, I can also just check all the fields. I am confident  a.String() == b.String() iff a == b since the function is generated based on all of the fields [here](https://github.com/kubernetes/apimachinery/blob/v0.28.1/pkg/apis/meta/v1/generated.pb.go#L4610-L4624)

Reply:
I'd go with the value comparison, that seems more solid to me than the string comparison.

Reply:
In case it can be useful, controller-runtimes implements a `SetOwnerReference()` function: https://github.com/kubernetes-sigs/controller-runtime/blob/main/pkg/controller/controllerutil/controllerutil.go#L97. If you follow the code you see that they compare the group, the kind and the name. Not sure if we can use that function directly, maybe you know better here.